pyfr/integrators/dual/pseudo/pseudocontrollers.py

```diff
@@ -90,3 +97,120 @@ def pseudo_advance(self, tcurr, tout, dt):
+
+        sord = self._pseudo_stepper_order
+        expa = alpha / sord
+        expb = beta / sord
+
+        # Limits for the local pseudo-time-step size
+        dtau_min = self._dtau
+        dtau_max = dtau_maxf * dtau_min
+
+        # Template parameters
+        tplargs = dict(ndims=self.system.ndims, nvars=self.system.nvars,
+                       expa=expa, expb=expb, maxf=maxf, minf=minf, saff=saff,
+                       sord=sord, atol=atol, dtau_min=dtau_min,
+                       dtau_max=dtau_max)
+
+        # Register a kernel to compute local error
+        self.backend.pointwise.register(
+            'pyfr.integrators.dual.pseudo.kernels.localerrest'
+        )
+
+        self.err_prev = {}
+
+        for (ename, ele), shape in zip(self.system.ele_map.items(),
+                                       self.system.ele_shapes):
+            # Allocate storage for previous error
+            self.err_prev[ename] = self.backend.matrix(
+                shape, np.ones(shape), tags={'align'}
+            )
+
+            # Append the error kernels to the proxylist
+            self.pintgkernels['eles', 'localerrest'].append(
+                self.backend.kernel(
+                    'localerrest', tplargs=tplargs,
+                    dims=[ele.nupts, ele.neles], err=ele.scal_upts_inb,
+                    errprev=self.err_prev[ename],
+                    dtau_upts=self.dtau_upts[ename]
+                )
+            )
+
+        self.backend.commit()
+
+    @property
+    def _pseudo_controller_needs_lerrest(self):
+        return True
+
+    def localerrest(self, errbank):
+        self.system.eles_scal_upts_inb.active = errbank
+        self._queue % self.pintgkernels['eles', 'localerrest']()
+
+    def conv_mon(self, i, minniters, dtau=None):
+        if i >= minniters - 1:
+            # Subtract the current solution from the previous solution
+            self._add(-1.0, self._idxprev, 1.0, self._idxcurr)
+
+            # Divide by 1/dtau
+            self.localdtau(self._idxprev, inv=1)
+
+            # Reduction
+            resid = tuple(self._resid(1.0, self._idxprev))
+        else:
+            resid = None
+
+        # Increment the step count
+        self.npseudosteps += 1
+
+        self.pseudostepinfo.append((self.npseudosteps, i + 1, resid))
+
+        converged = all(
+            r <= t for r, t in zip(resid, self._pseudo_residtol.values())
+        )
+
+        return resid and converged
+
+    def pseudo_advance(self, tcurr, dt_lmtr, dt):
+        self.tcurr = tcurr
+
+        # Pseudo time-step limiter close to outputs
+        self.dtau_lmtr = dt_lmtr
+
+        for i in range(self.maxniters):
+            # Take the step
+            self._idxcurr, self._idxprev, self._idxerr = self.step(self.tcurr,
+                                                                   dt)
+            if self.dtau_lmtr > self._dtau:
+                self.localerrest(self._idxerr)
```

Comment:
We have localerrest and conv_mon, lets pick a consistent style (I have no particular preference either way).